### PR TITLE
Bounds checking for RAJA views.

### PR DESCRIPTION
**Pull request by artv3**
[comment]: # (#################################################################)
[comment]: # (Copyright 2016-19, Lawrence Livermore National Security, LLC)
[comment]: # (and RAJA project contributors. See the RAJA/COPYRIGHT file)
[comment]: # (for details.)
[comment]: #
[comment]: # (# SPDX-License-Identifier: BSD-3-Clause)
[comment]: # (#################################################################)

# Summary

- This PR is adds bounds checking to RAJA views.
- It does the following (modify list as needed):
  Adds a run time check to make sure indices are within bounds of RAJA array dimensions.

### CODE REVIEW

**Comment by artv3** [comment]
One issue here is that I am using asserts to check for out of bounds indexing. If users have the NDEBUG macro defined then asserts will do nothing. Is there an alternative to using asserts that would also work with CUDA?

File: include/RAJA/util/Layout.hpp [hunk 1]
```diff
@@ -174,10 +190,18 @@ struct LayoutBase_impl<camp::idx_seq<RangeInts...>, IdxLin, StrideOneDim> {
...
+#if defined(BOUNDS_CHECK)
+  RAJA_INLINE RAJA_HOST_DEVICE IdxLin operator()(
+      Indices... indices) const
+  {
+    BoundsCheck<0>(indices...);
+#else
   RAJA_INLINE RAJA_HOST_DEVICE constexpr IdxLin operator()(
       Indices... indices) const
   {
+#endif
```
Comment (MrBurmark) [inline include/RAJA/util/Layout.hpp]: Would it make more sense to have a BOUNDS_CHECK_constexpr instead of repeating this?

Comment (artv3) [reply include/RAJA/util/Layout.hpp]: Ah, are you saying define BOUNDS_CHECK_constexp as a macro? 
#defined BOUNDS_CHECK_constexp constexpr ?

Comment (MrBurmark) [reply include/RAJA/util/Layout.hpp]: yes, defined as constexpr if BOUNDS_CHECK isn't defined, and nothing if it is.

**Comment by artv3** [comment]
I added a RAJA_ASSERT macro that changes behavior if NDEBUG is present or not. 
```
#if !defined(NDEBUG)
#define RAJA_ASSERT(EXP) assert( (EXP) )
#endif

#if defined(NDEBUG)
#define RAJA_ASSERT(EXP) asm ("trap;")
#endif
```

This seems to do the trick.

File: include/RAJA/config.hpp.in [hunk 2]
```diff
@@ -178,6 +187,25 @@ const int DATA_ALIGN = @RAJA_DATA_ALIGN@;
...
+#if defined(ENABLE_BOUNDS_CHECK)
+#define BOUNDS_CHECK
+#define BOUNDS_CHECK_constexpr
+
+#if !defined(NDEBUG)
+#define RAJA_ASSERT(EXP) assert( (EXP) )
+#endif
+
+#if defined(NDEBUG)
+#define RAJA_ASSERT(EXP) asm ("trap;")
```
Comment (ajkunen) [inline include/RAJA/config.hpp.in]: how portable is this? will it work with CUDA/HIP and things like Arm or POWER?

Comment (artv3) [reply include/RAJA/config.hpp.in]: I've had success with CUDA and power, but I haven't tried on Arm. I can search around to double check.

Comment (artv3) [reply include/RAJA/config.hpp.in]: An option is to use abort (standard C) for anything on the host and asm for cuda device code. For Hip...i'm not too sure...

File: CMakeLists.txt [hunk 3]
```diff
@@ -63,6 +63,7 @@ option(ENABLE_COVERAGE "Enable coverage (only supported with GCC)" Off)
 option(ENABLE_FORCEINLINE_RECURSIVE "Enable Forceinline recursive (only supported with Intel compilers)" On)
 option(ENABLE_BENCHMARKS "Build benchmarks" Off)
 option(RAJA_DEPRECATED_TESTS "Test deprecated features" Off)
+option(ENABLE_BOUNDS_CHECK "Enable bounds checking in RAJA::Views" Off)
```
Comment (rhornung67) [inline CMakeLists.txt]: I think we should  be explicit about which CMake variables are specific to RAJA and which are not; e.g., include a prefix such as "RAJA_" on those that our ours. So, I would prefer this to be RAJA_ENABLE_BOUNDS_CHECK or similar. Also, we should be consistent in the names of these variables; e.g., RAJA_DEPRECATED_TESTS should be RAJA_ENABLE_DEPRECATED_TESTS. @artv3 @davidbeckingsale @MrBurmark @ajkunen what do you think?

Comment (ajkunen) [reply CMakeLists.txt]: I agree

File: include/RAJA/util/Layout.hpp [hunk 4]
```diff
@@ -166,6 +167,28 @@ struct LayoutBase_impl<camp::idx_seq<RangeInts...>, IdxLin, StrideOneDim> {
   {
   }
 
+  /*!
+   * Methods to performs bounds checking in layout objects
+   */
+  template<camp::idx_t N, typename Idx>
+  RAJA_INLINE RAJA_HOST_DEVICE void BoundsCheckError(Idx idx) const
+  {
+    printf("View index %ld is out of bounds [0, %ld) \n" ,static_cast<long int>(idx), static_cast<long int>(sizes[N]));
```
Comment (ajkunen) [inline include/RAJA/util/Layout.hpp]: Two things:  this is a Layout not a View so the comment should reflect that (people use Layouts without Views... or at least I do), and how does this work with Layouts that don't have a [0,N) range? such as the OffsetLayout?

File: CMakeLists.txt (hunk 3 above)
Comment (rhornung67) [reply CMakeLists.txt]: @artv3 let's address the CMake variable name change on a different PR. I'll discuss with you.

Comment (artv3) [reply CMakeLists.txt]: I like that idea, I can change that in this PR.

File: include/RAJA/util/Layout.hpp (hunk 4 above)
Comment (artv3) [reply include/RAJA/util/Layout.hpp]: That's a good point, I'll fix the comment. Offsets layouts actually go through regular views to do the index calculation. The offset layouts just have an additional step where they have to remove the "offsets". I can do something similar for the offset layouts and intercept right before the offsets are removed to get more accurate information about the interval range.

Comment (artv3) [reply include/RAJA/util/Layout.hpp]: Done.

File: include/RAJA/config.hpp.in (hunk 2 above)
Comment (artv3) [reply include/RAJA/config.hpp.in]: @ajkunen, for AMD gpus we can use 
```
      asm("s_trap 2");
 
aborts the kernel with a message to stdout:
 
### HCC STATUS_CHECK Error: HSA_STATUS_ERROR_EXCEPTION (0x1016) at file:mcwamp_hsa.cpp line:1194
Aborted (core dumped)
```

**Comment by artv3** [comment]
I think this branch is ready for a review, may I kindly get a review?

File: include/RAJA/config.hpp.in [hunk 5]
```diff
@@ -178,6 +187,29 @@ const int DATA_ALIGN = @RAJA_DATA_ALIGN@;
...
+#define RAJA_ASSERT(EXP) assert( (EXP) )
+#endif
+
+#if defined(NDEBUG) && defined(__CUDA_ARCH__)
+#define RAJA_ASSERT(EXP) asm ("trap;")
+#endif
+
+#if defined(NDEBUG)
+#define RAJA_ASSERT(EXP) abort();
+#endif
```
Comment (MrBurmark) [inline include/RAJA/config.hpp.in]: Won't this second ifdef always redefine RAJA_ASSERT in device code.

File: CMakeLists.txt [hunk 6]
```diff
@@ -63,6 +63,7 @@ option(ENABLE_COVERAGE "Enable coverage (only supported with GCC)" Off)
 option(ENABLE_FORCEINLINE_RECURSIVE "Enable Forceinline recursive (only supported with Intel compilers)" On)
 option(ENABLE_BENCHMARKS "Build benchmarks" Off)
 option(RAJA_DEPRECATED_TESTS "Test deprecated features" Off)
+option(ENABLE_BOUNDS_CHECK "Enable bounds checking in RAJA::Views/Layouts" Off)
```
Comment (rhornung67) [inline CMakeLists.txt]: Can we change this to RAJA_ENABLE_BOUNDS_CHECK for this PR?

File: docs/sphinx/user_guide/config_options.rst [hunk 7]
```diff
@@ -98,6 +98,14 @@ and their default settings:
       =========================   ======================
       ENABLE_WARNINGS_AS_ERRORS   Off
       =========================   ======================
+
+     RAJA Views/Layouts may be configured to check for out of bounds 
+     indexing:
+      =========================   ======================
+      Variable                    Default
+      =========================   ======================
+      ENABLE_BOUNDS_CHECK         Off
```
Comment (rhornung67) [inline docs/sphinx/user_guide/config_options.rst]: Change this too, to reflect CMake variable name change.

File: docs/sphinx/user_guide/feature/view.rst [hunk 8]
```diff
@@ -285,3 +285,14 @@ will always return zero for each dimension with zero extent. For example::
    // The inverse mapping always produces a 0 for j
    int i,j,k;
    layout.toIndices(lin2, i, j, k); // i,j,k = {0, 0, 1}
+
+
+------------------------------------
+RAJA View/Layouts Bounds Checking
+------------------------------------
+
+The RAJA CMake variable ``ENABLE_BOUNDS_CHECK`` may be used to turn on/off
```
Comment (rhornung67) [inline docs/sphinx/user_guide/feature/view.rst]: This one too.

File: include/RAJA/config.hpp.in [hunk 9]
```diff
@@ -68,6 +68,15 @@
...
 #cmakedefine ENABLE_FORCEINLINE_RECURSIVE
 
+/*!
+ ******************************************************************************
+ *
+ * \brief Add bounds check to views and layouts
+ *
+ ******************************************************************************
+ */
+#cmakedefine ENABLE_BOUNDS_CHECK
```
Comment (rhornung67) [inline include/RAJA/config.hpp.in]: Another one.

File: include/RAJA/config.hpp.in [hunk 10]
```diff
@@ -178,6 +187,33 @@ const int DATA_ALIGN = @RAJA_DATA_ALIGN@;
 #define RAJA_RESTRICT __restrict__
 #endif
 
+//
+// Runtime bounds checking for Views
+//
+#if defined(ENABLE_BOUNDS_CHECK)
```
Comment (rhornung67) [inline include/RAJA/config.hpp.in]: Here's another.

File: include/RAJA/config.hpp.in [hunk 11]
```diff
@@ -178,6 +187,33 @@ const int DATA_ALIGN = @RAJA_DATA_ALIGN@;
...
+#define BOUNDS_CHECK
```
Comment (rhornung67) [inline include/RAJA/config.hpp.in]: This should probably be changed to RAJA_BOUNDS_CHECK_INTERNAL (or something similarly unique) so we don't conflict with other project's macro definitions.

File: CMakeLists.txt (hunk 6 above)
Comment (artv3) [reply CMakeLists.txt]: Sure thing!

**Pull request merged**